ProjectVoyage/Gateways/MessageGateway.m

```diff
@@ -0,0 +1,124 @@
+//
+//  MessageGateway.m
+//  Whatsapp
+//
+//  Created by Rafael Castro on 7/4/15.
+//  Copyright (c) 2015 HummingBird. All rights reserved.
+//
+
+#import "MessageGateway.h"
+#import "LocalStorage.h"
+
+@interface MessageGateway()
+@property (strong, nonatomic) NSMutableArray *messages_to_send;
+@end
+
+
+@implementation MessageGateway
+
++(id)sharedInstance
+{
+    static MessageGateway *sharedInstance = nil;
+    static dispatch_once_t onceToken;
+    dispatch_once(&onceToken, ^{
+        sharedInstance = [[self alloc] init];
+    });
+    return sharedInstance;
+}
+-(id)init
+{
+    self = [super init];
+    if (self)
+    {
+        self.messages_to_send = [[NSMutableArray alloc] init];
+    }
+    return self;
+}
+-(void)loadOldMessages
+{
+    NSArray *array = [[LocalStorage sharedInstance] queryMessagesForChatID:self.chat.identifier];
+    if (self.delegate)
+    {
+        [self.delegate gatewayDidReceiveMessages:array];
+    }
+    NSArray *unreadMessages = [self queryUnreadMessagesInArray:array];
+    [self updateStatusToReadInArray:unreadMessages];
+}
+-(void)updateStatusToReadInArray:(NSArray *)unreadMessages
+{
+    NSMutableArray *read_ids = [[NSMutableArray alloc] init];
+    for (Message *message in unreadMessages)
+    {
+        message.status = MessageStatusRead;
+        [read_ids addObject:message.identifier];
+    }
+    self.chat.numberOfUnreadMessages = 0;
+    [self sendReadStatusToMessages:read_ids];
+}
+-(NSArray *)queryUnreadMessagesInArray:(NSArray *)array
+{
+    NSPredicate *predicate = [NSPredicate predicateWithFormat:@"SELF.status == %d", MessageStatusReceived];
+    return [array filteredArrayUsingPredicate:predicate];
+}
+-(void)news
+{
+    
+}
+-(void)dismiss
+{
+    self.delegate = nil;
+}
+
+-(void)fakeMessageUpdate:(Message *)message
+{
+    [self performSelector:@selector(updateMessageStatus:) withObject:message afterDelay:2.0];
+}
+-(void)updateMessageStatus:(Message *)message
+{
+    if (message.status == MessageStatusSending)
```

Comment:
When you have a value like `status` with a specific set of values it could be, and you and you want to perform a different action for each value I find it better to use a switch statement. The primary benefit is that a long if, else if state doesn't have type safety to have handled all possible message status values, but a switch case will throw a compile time error if you forgot to handle any of them.